File: 095. Unique Binary Search Trees II/Java/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<TreeNode> generateTrees(int n) {
        class Utils {
            List<TreeNode> generate(int start, int end) {
                List<TreeNode> result = new ArrayList<>();
                if (start > end) {
                    result.add(null);
                    return result;
                }
                for (int i = start; i <= end; i++) {
                    List<TreeNode> lefts = generate(start, i - 1);
                    List<TreeNode> rights = generate(i + 1, end);
                    for (TreeNode left : lefts) {
                        for (TreeNode right : rights) {
                            TreeNode root = new TreeNode(i);
                            root.left = left;
                            root.right = right;
                            result.add(root);
                        }
                    }
                }
                return result;
            }
        }
        if (n == 0) {
            return new ArrayList<>();
        }
        Utils utils = new Utils();
        return utils.generate(1, n);
    }
}

File: 095. Unique Binary Search Trees II/C++/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<TreeNode *> generateTrees(int n) {
        class Utils {
        public:
            vector<TreeNode *> generate(int start, int end) {
                vector<TreeNode *> result;
                if (start > end) {
                    result.push_back(nullptr);
                    return result;
                }
                for (int i = start; i <= end; i++) {
                    vector<TreeNode *> lefts = generate(start, i - 1);
                    vector<TreeNode *> rights = generate(i + 1, end);
                    for (TreeNode *left : lefts) {
                        for (TreeNode *right : rights) {
                            TreeNode *root = new TreeNode(i);
                            root->left = left;
                            root->right = right;
                            result.push_back(root);
                        }
                    }
                }
                return result;
            }
        };
        if (n == 0) {
            return vector<TreeNode *>();
        }
        Utils utils;
        return utils.generate(1, n);
    }
};